src/overlay/OverlayEnvelopeHandler.cpp

```diff
@@ -0,0 +1,201 @@
+    return status;
+}
+
+EnvelopeHandler::EnvelopeStatus
+OverlayEnvelopeHandler::processEnvelope(Peer::pointer peer,
+                                        SCPEnvelope const& envelope)
+{
+    if (mApp.getConfig().MANUAL_CLOSE)
+    {
+        return EnvelopeHandler::ENVELOPE_STATUS_DISCARDED;
+    }
+
+    if (Logging::logDebug("Overlay"))
+        CLOG(DEBUG, "Overlay")
+            << "recvSCPEnvelope"
+            << " from: "
+            << mApp.getConfig().toShortString(envelope.statement.nodeID)
+            << " s:" << envelope.statement.pledges.type()
+            << " i:" << envelope.statement.slotIndex
+            << " a:" << mApp.getStateHuman();
+
+    if (envelope.statement.nodeID == mLocalNodeID)
+    {
+        CLOG(DEBUG, "Overlay") << "recvSCPEnvelope: skipping own message";
+        return EnvelopeHandler::ENVELOPE_STATUS_DISCARDED;
+    }
+
+    // If envelopes are out of our validity brackets, we just ignore them.
+    if (envelope.statement.slotIndex > mMax ||
+        envelope.statement.slotIndex < mMin)
+    {
+        CLOG(DEBUG, "Overlay") << "Ignoring SCPEnvelope outside of range: "
+                               << envelope.statement.slotIndex << "( " << mMin
+                               << "," << mMax << ")";
+        return EnvelopeHandler::ENVELOPE_STATUS_DISCARDED;
+    }
+
+    auto const& nodeID = envelope.statement.nodeID;
+    if (!mApp.getHerder().isNodeInQuorum(nodeID))
+    {
+        CLOG(DEBUG, "Herder")
+            << "Dropping envelope from "
+            << mApp.getConfig().toShortString(nodeID) << " (not in quorum)";
+        return EnvelopeHandler::ENVELOPE_STATUS_DISCARDED;
+    }
+
+    auto status = mApp.getOverlayManager().getPendingEnvelopes().handleEnvelope(
+        nullptr, envelope);
+    if (status != EnvelopeHandler::ENVELOPE_STATUS_READY)
+    {
+        return status;
+    }
+
+    if (mApp.getHerder().seen(envelope))
+    {
+        return EnvelopeHandler::ENVELOPE_STATUS_PROCESSED;
+    }
+
+    mApp.getHerder().push(envelope);
+    return EnvelopeHandler::ENVELOPE_STATUS_READY;
+}
+
+void
+OverlayEnvelopeHandler::getQuorumSet(Peer::pointer peer, Hash const& hash)
+{
+    if (auto qset = mApp.getItemFetcher().getQuorumSet(hash))
+    {
+        peer->sendSCPQuorumSet(qset);
+    }
+    else
+    {
+        if (Logging::logTrace("Overlay"))
+            CLOG(TRACE, "Overlay") << "No quorum set: " << hexAbbrev(hash);
+        peer->sendDontHave(SCP_QUORUMSET, hash);
+        // do we want to ask other people for it?
+    }
+}
+
+std::set<SCPEnvelope>
+OverlayEnvelopeHandler::quorumSet(Peer::pointer peer, SCPQuorumSet const& qSet,
+                                  bool force)
+{
+    auto result =
+        mApp.getOverlayManager().getPendingEnvelopes().handleQuorumSet(qSet,
+                                                                       force);
+    if (result.first)
+    {
+        mApp.getHerder().clearNodesInQuorumCache();
```

Comment:
not sure I like that we're clearing the cache here - maybe we don't have the choice yet (until we properly manage LRU data in overlay vs "real" data in herder)

Reply:
Looks like this is not needed. I've removed that from this branch and it passes all the tests.
I've also removed corresponding line from master and it also passes all the tests.